craft_application/application.py

```diff
@@ -375,17 +370,18 @@ def is_managed(self) -> bool:
 
     def run_managed(self, platform: str | None, build_for: str | None) -> None:
         """Run the application in a managed instance."""
-        if not self._build_plan:
+        build_planner = self.services.get("build_plan")
+        if platform:
+            build_planner.set_platforms(platform)
+        if build_for:
+            build_planner.set_build_fors(build_for)
+        plan = build_planner.plan()
+
+        if not plan:
             raise errors.EmptyBuildPlanError
 
         extra_args: dict[str, Any] = {}
-        for build_info in self._build_plan:
-            if platform and platform != build_info.platform:
-                continue
-
-            if build_for and build_for != build_info.build_for:
-                continue
-
+        for build_info in plan:
```

Comment:
Since I'm not going to be in until Thursday I'll leave a higher-level comment here: I agree with the overall direction that the PR heads (most of my comments are suggestions), but I'm not seeing how this enables hybrid commands. Is the upcoming PR going to re-work this whole `run_managed()` bit?

Reply:
If you have a satisfying answer to this question^, you can consider this PR approved by me *in absentia* ;)

Reply:
The next PR is going to move the responsibility of iterating over the build plan and running in various managed instances onto the command, eliminating `run_managed` (though replacing it with a convenience method on the `ProviderService`). This and the previous PR needed to happen to detangle everything that was overloading `run_managed`.